search/searcher/search_phrase.go

```diff
@@ -150,49 +150,68 @@ func (s *PhraseSearcher) checkCurrMustMatch(ctx *search.SearchContext) *search.D
-	INNER:
-		for i := 1; i < len(s.terms); i++ {
-			nextTerm := s.terms[i]
-			if nextTerm == "" {
-				continue
-			}
-			// look through all these term locations
-			// to try and find the correct offsets
-			nextLocations, ok := termLocMap[nextTerm]
-			if !ok {
-				continue OUTER
-			}
-			for _, nextLocation := range nextLocations {
-				if nextLocation.Pos == location.Pos+float64(i) && nextLocation.SameArrayElement(location) {
-					// found a location match for this term
-					crvtlm.AddLocation(nextTerm, nextLocation)
-					continue INNER
-				}
-			}
-			// if we got here we didn't find a location match for this term
-			continue OUTER
-		}
-		// if we got here all the terms matched
-		freq++
-		search.MergeTermLocationMaps(rvtlm, crvtlm)
+func (s *PhraseSearcher) checkCurrMustMatchField(ctx *search.SearchContext, tlm search.TermLocationMap) (int, search.TermLocationMap) {
+	paths := findPhrasePaths(nil, s.terms, tlm, 1, nil)
+	rv := make(search.TermLocationMap, 0)
+	for _, p := range paths {
+		p.MergeInto(rv)
+	}
+	return len(paths), rv
+}
+
+type phrasePart struct {
+	term string
+	loc  *search.Location
+}
+
+type phrasePath []*phrasePart
+
+func (p phrasePath) MergeInto(in search.TermLocationMap) {
+	for _, pp := range p {
+		in[pp.term] = append(in[pp.term], pp.loc)
 	}
+}
 
-	return freq, rvtlm
+// findPhrasePaths is a function to identify phase matches from a set of known
+// term locations.  the implementation is recursive, so care must be taken
+// with arguments and return values.
+//
+// prev - the previous location, nil on first invocation
+// phraseTerms - slice containing the phrase terms themselves
+//               may contain empty string as placeholder (don't care)
+// tlm - the Term Location Map containing all relevant term locations
+// offset - the offset from the previous that this next term must match
+// p - the current path being explored (appended to in recursive calls)
+//     this is the primary state being built during the traversal
+//
+// returns slice of paths, or nil if invocation did not find any successul paths
+func findPhrasePaths(prev *search.Location, phraseTerms []string, tlm search.TermLocationMap, offset int, p phrasePath) []phrasePath {
+
+	// no more terms
+	if len(phraseTerms) < 1 {
+		return []phrasePath{p}
+	}
+
+	car := phraseTerms[0]
+	cdr := phraseTerms[1:]
+
+	// empty term is treated as match (continue), but offset now +1
+	if car == "" {
+		return findPhrasePaths(prev, cdr, tlm, offset+1, p)
+	}
+
+	// locations for this term
+	locations := tlm[car]
+	var rv []phrasePath
+	for _, loc := range locations {
+		// check each location against prev (nil treated as match, the initial case)
+		if prev == nil || (prev.Pos+float64(offset) == loc.Pos && prev.SameArrayElement(loc)) {
+			// this location works, add it to the path (but not for empty term)
+			px := append(p, &phrasePart{term: car, loc: loc})
+			rv = append(rv, findPhrasePaths(loc, cdr, tlm, 1, px)...)
```

Comment:
I actually think this part of the code will change if I add support for slop, because then there will be like a total slop value that can be used up.  And each match uses up some slop (or not if its in the exact place you expect), until you run out of slop.